### ISSUE TITLE
High CPU due to lots of `git ls-files`

### ISSUE THREAD

**Issue by xiaochuanyu** [lewis6991/gitsigns.nvim#620]
### Description

**TLDR**: `gitsigns` causes constant high CPU usage for me due to spamming `git ls-files` in a large repo.

I work in a very large repo where `git ls-files` consumes a lot of CPU and take about 2 seconds to run.
`gitsigns` seem to spawn  `git ls-files` constantly for every buffer and may spawn multiple for each file.
For example I saw 3-4 identical `git ls-files` for the same file running constantly if I run `ps | grep some_file`:
```
git --no-pager --literal-pathspecs --git-dir /Users/xcyu/big-repo/.git -c core.quotepath=off ls-files --stage --others --exclude-standard --eol /Users/xcyu/big-repo/some_file
```
I compared this with Visual Studio Code which seems to only run `git ls-files --stage -- /Users/xcyu/big-repo/some_file` (from repo root as working directory) maybe every 3-5s when a particular file is visible in UI.

If I open 5+ buffers I get 100% CPU and eventually error messages in vim with `Error: EMFILE, too many open files` when additional `git ls-files` are spawned by `gitsigns`.

### Neovim version
```
NVIM v0.7.2 Build type: Release LuaJIT 2.1.0-beta3 Compiled by brew@HMBRW-A-001-M1-004.local  Features: +acl +iconv +tui See ":help feature-compile"     system vimrc file: "$VIM/sysinit.vim"   fall-back for $VIM: "/opt/homebrew/Cellar/neovim/0.7.2_1/share/nvim"  Run :checkhealth for more info
```
### Operating system and version

macOS Monterey 12.5.1 (21G83)

### Expected behavior

Not constant 100% CPU with 5+ buffers.

### Actual behavior

Constant 100% CPU with 5+ buffers.

### Minimal config

```Lua
vim.o.packpath = '/tmp/nvim/site'

local plugins = {
  gitsigns = 'https://github.com/lewis6991/gitsigns.nvim',
  -- ADD OTHER PLUGINS _NECESSARY_ TO REPRODUCE THE ISSUE
}

for name, url in pairs(plugins) do
  local install_path = '/tmp/nvim/site/pack/test/start/'..name
  if vim.fn.isdirectory(install_path) == 0 then
    vim.fn.system { 'git', 'clone', '--depth=1', url, install_path }
  end
end

require('gitsigns').setup{
  debug_mode = true, -- You must add this to enable debug messages
  update_debounce = 10000,
}

-- ADD INIT.LUA SETTINGS THAT IS _NECESSARY_ FOR REPRODUCING THE ISSUE
```

### Steps to reproduce

1. `nvim --clean -u minimal.lua`
2. Open a file in large repo.

### Gitsigns debug messages
```
signs.init: Using vimfn signs
run_job: git --no-pager --literal-pathspecs --version
fn: Highlight GitSignsAdd is already defined
fn: Highlight GitSignsChange is already defined
fn: Highlight GitSignsDelete is already defined
fn: Deriving GitSignsAddNr from GitSignsAdd
fn: Deriving GitSignsChangeNr from GitSignsChange
fn: Deriving GitSignsDeleteNr from GitSignsDelete
fn: Deriving GitSignsAddLn from DiffAdd
fn: Deriving GitSignsChangeLn from DiffChange
fn: Deriving GitSignsAddPreview from DiffAdd
fn: Deriving GitSignsDeletePreview from DiffDelete
fn: Deriving GitSignsCurrentLineBlame from NonText
fn: Deriving GitSignsAddInline from TermCursor
fn: Deriving GitSignsDeleteInline from TermCursor
fn: Deriving GitSignsChangeInline from TermCursor
fn: Deriving GitSignsAddLnInline from GitSignsAddInline
fn: Deriving GitSignsChangeLnInline from GitSignsChangeInline
fn: Deriving GitSignsDeleteLnInline from GitSignsDeleteInline
fn: Deriving GitSignsAddLnVirtLn from GitSignsAddLn
fn: Deriving GitSignsChangeVirtLn from GitSignsChangeLn
fn: Deriving GitSignsDeleteVirtLn from DiffDelete
fn: Deriving GitSignsAddLnVirtLnInLine from GitSignsAddLnInline
fn: Deriving GitSignsChangeVirtLnInLine from GitSignsChangeLnInline
fn: Deriving GitSignsDeleteVirtLnInLine from GitSignsDeleteLnInline
run_job: git --no-pager --literal-pathspecs rev-parse --show-toplevel --absolute-git-dir --abbrev-ref HEAD
attach(3): Attaching (trigger=BufRead)
run_job: git --no-pager --literal-pathspecs config user.name
run_job: git --no-pager --literal-pathspecs rev-parse --show-toplevel --absolute-git-dir --abbrev-ref HEAD
run_job: git --no-pager --literal-pathspecs --git-dir /Users/xcyu/big-repo/.git -c core.quotepath=off ls-files --stage --others --exclude-standard --eol /Users/xcyu/big-repo/some_file
watch_gitdir(3): Watching git dir
run_job: git --no-pager --literal-pathspecs --git-dir /Users/xcyu/big-repo/.git show :0:some_file
watcher_cb(3): Git dir update
run_job: git --no-pager --literal-pathspecs rev-parse --show-toplevel --absolute-git-dir --abbrev-ref HEAD
run_job: git --no-pager --literal-pathspecs --git-dir /Users/xcyu/big-repo/.git -c core.quotepath=off ls-files --stage --others --exclude-standard --eol /Users/xcyu/big-repo/some_file
watcher_cb(3): Git dir update
run_job: git --no-pager --literal-pathspecs rev-parse --show-toplevel --absolute-git-dir --abbrev-ref HEAD
run_job: git --no-pager --literal-pathspecs --git-dir /Users/xcyu/big-repo/.git -c core.quotepath=off ls-files --stage --others --exclude-standard --eol /Users/xcyu/big-repo/some_file
update(3): updates: 1, jobs: 10
watcher_cb(3): Git dir update
run_job: git --no-pager --literal-pathspecs rev-parse --show-toplevel --absolute-git-dir --abbrev-ref HEAD
run_job: git --no-pager --literal-pathspecs --git-dir /Users/xcyu/big-repo/.git -c core.quotepath=off ls-files --stage --others --exclude-standard --eol /Users/xcyu/big-repo/some_file
watcher_cb(3): File not changed
watcher_cb(3): Git dir update
watcher_cb(3): File not changed
run_job: git --no-pager --literal-pathspecs rev-parse --show-toplevel --absolute-git-dir --abbrev-ref HEAD
run_job: git --no-pager --literal-pathspecs --git-dir /Users/xcyu/big-repo/.git -c core.quotepath=off ls-files --stage --others --exclude-standard --eol /Users/xcyu/big-repo/some_file
watcher_cb(3): Git dir update
run_job: git --no-pager --literal-pathspecs rev-parse --show-toplevel --absolute-git-dir --abbrev-ref HEAD
run_job: git --no-pager --literal-pathspecs --git-dir /Users/xcyu/big-repo/.git -c core.quotepath=off ls-files --stage --others --exclude-standard --eol /Users/xcyu/big-repo/some_file
watcher_cb(3): File not changed
watcher_cb(3): Git dir update
...last 4 lines repeated forever
```

**Comment by xiaochuanyu** [comment]
K I updated my config to disable `watch_gitdir` and I think its got a little better:
```lua
gitsigns.setup {
  debug_mode = true,
  watch_gitdir = {
    enabled = false,
  },
  update_debounce = 100,
  current_line_blame = false,
}
```
Just looking at running processes, I see much less `ls-files` running and only 1 instance per 2-3 sec.
My debug messages look like the following repeated every ~2 sec:
```
...
watcher_cb(8): File not changed
watcher_cb(8): Git dir update
run_job: git --no-pager --literal-pathspecs rev-parse --show-toplevel --absolute-git-dir --abbrev-ref HEAD
run_job: git --no-pager --literal-pathspecs --git-dir /Users/xcyu/big-repo/.git -c core.quotepath=off ls-files --stage --others --exclude-standard --eol /Users/xcyu/big-repo/some_file
...
```
In terms of setup, I just load up vim in that repo & open `some_file` open (it has unstaged changes if that matters).
Is this expected behavior or is something still wrong?

**Comment by lewis6991** [comment]
With `watch_gitdir.enabled == false` you shouldn't be seeing any `watcher_cb` entries in the log.

Can you confirm you are using the latest Gitsigns?

**Comment by xiaochuanyu** [comment]
Ok I think that really fixed. Thank you so much!
